### DISCUSSION TITLE
ros2_ament_setup to fix AMENT_PREFIX_PATH not set?

### DISCUSSION THREAD

**Discussion by dbruggner** [mvukov/rules_ros2#36] [General]
I'm working on integrating image_common into rules_ros2: https://github.com/mvukov/rules_ros2/compare/main...dbruggner:rules_ros2:feature/add-image-common

I have a node which uses `image_transport`. When I start this node I get an error that the `AMENT_PREFIX_PATH` is not set. Now I have seen that you implemented a `ros2_ament_setup`. Is it possible to use this to set the `AMENT_PREFIX_PATH` for my node?

**Comment by ahans** [comment]
You can checkout the pluginlib [test](https://github.com/mvukov/rules_ros2/tree/main/ros2/test/pluginlib). You probably have to use `ros2_plugin` for a few files of `image_transport`. I'd expect in the regular colcon build, at least two shared objects are created, one for the default publisher and one for the default subscriber. Then there's [ros2_ament_setup](https://github.com/mvukov/rules_ros2/blob/main/ros2/ament.bzl#L138), which you can probably use to make the plugins into a "package". If I understand correctly, this rule creates a Python script with a function `set_up_ament()` that sets `AMENT_PREFIX_PATH` . For a C++ node, you would need to create a Python wrapper that first calls this function and then runs the binary. It could also make sense to extend that in a way that we also create a `cc_library` with such a function that you can then depend on wherever you need it.

**Comment by mvukov** [comment]
ros2_plugin takes a shared library and necessary info to create a plugins.xml-like file. Initial implementation is not general enough, I figured that out while working on feature/rosbag branch. In short, the current ros2_plugin assumes a single base class for all classes in a plugin, but more than one is fine as well. That rework should be done by the end of the week.

ros2_ament_setup at the moment collects all plugins defined in the dependency graph of the given deps and creates a file tree that ament_index_cpp can chew. I already upgraded ros2_test to use ros2_ament_setup under the hood. I still need to update ros2_launch. Since test&launch mechanisms are in Python, it makes sense to generate a simple Python setup like I do now. I envision that most folks would like to use ros2_launch|test instead of running bare nodes. In other words: if one would use ros2_test (soon ros2_launch) and there are plugins in the dep graph, all the plugins will be available automatically. This may be an overkill in some cases -- I am open to suggestions.

**Comment by mvukov** [comment]
While working on rosbag stuff, I had to rework ros2_plugin and ament setup. That's now in main. The rosbag stuff needs a bit more efforts as it heavily relies on runtime plugin loading.

**Reply by fernride-jissing** [reply]
Thanks for all the hard work, @mvukov ! The rules are pretty usable already and we'll be happy to contribute as soon as we worked ourselves a bit further into them.

I have been running into the same issue as mentioned in the thread title `AMENT_PREFIX_PATH not set`, but when running a `ros2_launch` target. The only way to make it work was by removing the `package` parameter from the launch description.

I wanted to drop the note here, since this is the thread that popped up when doing the Google search. I am aware it's not really addressing the issue at hand.

**Reply by mvukov** [reply]
Hi Jochen, I fixed issues with AMENT_PREFIX_PATH recently. It should work properly with rosl2_launch and ros2_test. Indeed, you have to remove `package` from launch description -- we don't need that with Bazel, just give a repo-relative path to the executable like demonstrated in examples. If you have an edge case, please provide more info.
